x-pack/plugins/security_solution/public/common/components/toolbar/bulk_actions/alert_bulk_tags.tsx

```diff
@@ -0,0 +1,170 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { EuiSelectableOption } from '@elastic/eui';
+import { EuiPopoverTitle, EuiSelectable, EuiButton } from '@elastic/eui';
+import type { TimelineItem } from '@kbn/timelines-plugin/common';
+import React, { memo, useCallback, useMemo, useState } from 'react';
+import { TAGS } from '@kbn/rule-data-utils';
+import type { EuiSelectableOnChangeEvent } from '@elastic/eui/src/components/selectable/selectable';
+import { getUpdateAlertsQuery } from '../../../../detections/components/alerts_table/actions';
+import { DEFAULT_ALERT_TAGS_KEY } from '../../../../../common/constants';
+import { useUiSetting$ } from '../../../lib/kibana';
+import { useSetAlertTags } from './use_set_alert_tags';
+import { useAppToasts } from '../../../hooks/use_app_toasts';
+import * as i18n from './translations';
+import { createInitialTagsState } from './helpers';
+
+interface BulkAlertTagsPanelComponentProps {
+  alertItems: TimelineItem[];
+  refetchQuery?: () => void;
+  setIsLoading: (isLoading: boolean) => void;
+  refresh?: () => void;
+  clearSelection?: () => void;
+  closePopoverMenu: () => void;
+}
+const BulkAlertTagsPanelComponent: React.FC<BulkAlertTagsPanelComponentProps> = ({
+  alertItems,
+  refresh,
+  refetchQuery,
+  setIsLoading,
+  clearSelection,
+  closePopoverMenu,
+}) => {
+  const [defaultAlertTagOptions] = useUiSetting$<string[]>(DEFAULT_ALERT_TAGS_KEY);
+  const { addSuccess, addError, addWarning } = useAppToasts();
+
+  const { setAlertTags } = useSetAlertTags();
+  const existingTags = useMemo(
+    () => alertItems.map((item) => item.data.find((data) => data.field === TAGS)?.value ?? []),
+    [alertItems]
+  );
+  const initalTagsState = useMemo(
+    () => createInitialTagsState(existingTags, defaultAlertTagOptions),
+    [existingTags, defaultAlertTagOptions]
+  );
+
+  const tagsToAdd: Record<string, boolean> = useMemo(() => ({}), []);
+  const tagsToRemove: Record<string, boolean> = useMemo(() => ({}), []);
+
+  const onUpdateSuccess = useCallback(
+    (updated: number, conflicts: number) => {
+      if (conflicts > 0) {
+        addWarning({
+          title: i18n.UPDATE_ALERT_TAGS_FAILED(conflicts),
+          text: i18n.UPDATE_ALERT_TAGS_FAILED_DETAILED(updated, conflicts),
+        });
+      } else {
+        addSuccess(i18n.UPDATE_ALERT_TAGS_SUCCESS_TOAST(updated));
+      }
+    },
+    [addSuccess, addWarning]
+  );
+
+  const onUpdateFailure = useCallback(
+    (error: Error) => {
+      addError(error.message, { title: i18n.UPDATE_ALERT_TAGS_FAILURE });
+    },
+    [addError]
+  );
+
+  const [selectableAlertTags, setSelectableAlertTags] =
+    useState<EuiSelectableOption[]>(initalTagsState);
+
+  const onTagsUpdate = useCallback(async () => {
+    closePopoverMenu();
+    const ids = alertItems.map((item) => item._id);
+    const query: Record<string, unknown> = getUpdateAlertsQuery(ids).query;
```

Comment:
If this is how you're typing the query, maybe `t.record(t.string, t.unknown)` is more accurate than the `t.object` used earlier?